pkg/image/cdi/download.go

```diff
@@ -1,19 +1,299 @@
+	downloadURL := fmt.Sprintf("http://%s.%s/images/%s", service.Name, vmImgNamespace, targetFileName)
+
+	// ensure the target file is ready
+	if err := wait.PollUntilContextTimeout(context.Background(), tickPolling, tickTimeout, true, func(context.Context) (bool, error) {
+		return cd.waitEndPointReady(downloadURL)
+	}); err != nil {
+		return fmt.Errorf("failed to wait for service ready")
+	}
+
+	downloadReq, err := http.NewRequestWithContext(req.Context(), http.MethodGet, downloadURL, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create the download request with VM Image(%s): %w", targetFileName, err)
+	}
+
+	downloadResp, err := cd.httpClient.Do(downloadReq)
+	if err != nil {
+		return fmt.Errorf("failed to send the download request with VM Image(%s): %w", targetFileName, err)
+	}
+	defer downloadResp.Body.Close()
+
+	rw.Header().Set("Content-Disposition", "attachment; filename="+targetFileName)
+	contentType := downloadResp.Header.Get("Content-Type")
+	if contentType != "" {
+		rw.Header().Set("Content-Type", contentType)
+	}
+
+	if _, err := io.Copy(rw, downloadResp.Body); err != nil {
+		return fmt.Errorf("failed to copy download content to target(%s), err: %w", targetFileName, err)
+	}
+
 	return nil
 }
+
+func (cd *Downloader) generateDownlodServerPod(podName, vmImageName string) *corev1.Pod {
+	virtImageStr := cd.getVirtHandlerImage()
+	clusterRepoImageStr := cd.getClusterRepoImage()
+
+	convertCmd := fmt.Sprintf("qemu-img convert -t none -T none -W -m 8 -f raw /tmp/image-vol -O qcow2 -c -S 4K /image-dir/%s.qcow2", vmImageName)
+
+	targetPod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: podName,
+			Labels: map[string]string{
+				"app": podName,
+			},
+		},
+		Spec: corev1.PodSpec{
+			Volumes: []corev1.Volume{
+				{
+					Name: "image-vol",
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: vmImageName,
+						},
+					},
+				},
+				{
+					Name: "image-dir",
+					VolumeSource: corev1.VolumeSource{
+						EmptyDir: &corev1.EmptyDirVolumeSource{},
+					},
+				},
+			},
+			InitContainers: []corev1.Container{
+				{
+					Name:            "image-coverter",
+					Image:           virtImageStr,
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					VolumeDevices: []corev1.VolumeDevice{
+						{
+							Name:       "image-vol",
+							DevicePath: "/tmp/image-vol",
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "image-dir",
+							MountPath: "/image-dir",
+						},
+					},
+					Command: []string{"/bin/sh", "-c"},
+					Args: []string{
+						convertCmd,
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:            "image-exporter",
+					Image:           clusterRepoImageStr,
+					ImagePullPolicy: corev1.PullIfNotPresent,
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: 80,
+							Protocol:      corev1.ProtocolTCP,
+						},
```

Comment:
We might need a separate ticket to discuss if this needs to be secured. E.g., using K8s [`TokenReview` API](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/token-review-v1/) to make sure only the controller's service account can invoke this endpoint.

Reply:
ditto

Reply:
Yeah, I thought you were correct. We could create another ticket to discuss authorization.